### ISSUE TITLE
ValueError: Edge 'Dangling_0' is not a dangling edge. This edge points to nodes: '__unnamed_node__' and '__unnamed_node__'

### ISSUE THREAD

**Issue by zyh1204** [google/TensorNetwork#395]
I wrote a code about a 3-order tensor contraction, but it goes wrong, Please help.
`
a = tn.Node(np.array([[[1, 2], [3, 4]], [[1, 2], [3, 4]]]))`
`edge1 = a[0] ^ a[1] ^ a[2]`
`res = tn.contract(edge)`

The mistake happens on the second line, any sugestions?

**Comment by MichaelMarien** [comment]
Hi, can you explain in more detail what you want to achieve? The error you see is thrown because `a[0] ^ a[1]` connects two of the 3 axis of the tensor `a`. The connected edge in between is not dangling anymore (no loose ends) and cannot be further connected to the third axis of the tensor, `a[2]`, which is the only dangling one remaining.

**Comment by Thenerdstation** [comment]
Correct. `a[0] ^ a[1]` creates a new non-dangling edge, so when you add `^ a[2]` on this line, it breaks with this error.

Though having edges like this isn't unheard of, it's just in this library we prefer to use `tn.CopyNode` instead. What you want looks like this.


```python
a = tn.Node(...)
cn = tn.CopyNode(2, 3)
for i in range (3):
  a[i] ^ cn[i]
tn.contract_copy_node(cn)
```


Though this does bring up a good point. Should we allow these auxiliary type edges? It would likely require another massive refector, but no API change.

**Comment by MichaelMarien** [comment]
Ok, now I see what the original intent was. Don't feel a big need to support this directly (without use of copy node) tbh.

**Comment by mganahl** [comment]
I'd also say no. This kind of contraction appears not so frequently.

**Comment by amilsted** [comment]
> "ValueError: Edge 'Dangling_0' is not a dangling edge. This edge points to nodes: '__unnamed_node__' and '__unnamed_node__'"

Just pointing out that, as error messages go, this is a pretty confusing 😄

**Comment by Thenerdstation** [comment]
Well adding names to the node would help. We could potentially add a "The node was defined here" debugging string like TF did, but from what I remember that caused op creation to be really slow

**Comment by amilsted** [comment]
I guess it's really the "Edge 'Dangling_0' is not a dangling edge" that gets me. Perhaps we shouldn't name edges "dangling" by default.

**Comment by Thenerdstation** [comment]
Oh yeah that's a good point. We should fix that...
